custom/ibc-transfer/keeper/msg_server.go

```diff
@@ -0,0 +1,82 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/ibc-go/v7/modules/apps/transfer/types"
+	ibctransfermiddlewaretypes "github.com/notional-labs/composable/v6/x/ibctransfermiddleware/types"
+)
+
+type msgServer struct {
+	Keeper
+	bank      types.BankKeeper
+	msgServer types.MsgServer
+}
+
+var _ types.MsgServer = msgServer{}
+
+func (k msgServer) Transfer(goCtx context.Context, msg *types.MsgTransfer) (*types.MsgTransferResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	params := k.Keeper.IbcTransfermiddleware.GetParams(ctx)
+	if params.ChannelFees != nil && len(params.ChannelFees) > 0 {
+		channelFee := findChannelParams(params.ChannelFees, msg.SourceChannel)
+		if channelFee != nil {
+			coin := findCoinByDenom(channelFee.AllowedTokens, msg.Token.Denom)
+			if coin != nil {
+				return &types.MsgTransferResponse{}, nil
+			}
+			minFee := coin.MinFee.Amount
+			charge := minFee
+			if charge.GT(msg.Token.Amount) {
+				charge = msg.Token.Amount
+			}
+
+			newAmount := msg.Token.Amount.Sub(charge)
+
+			if newAmount.IsPositive() {
+				percentageCharge := newAmount.QuoRaw(coin.Percentage)
+				newAmount = newAmount.Sub(percentageCharge)
+				charge = charge.Add(percentageCharge)
+			}
+
+			msgSender, err := sdk.AccAddressFromBech32(msg.Sender)
+			if err != nil {
+				return nil, err
+			}
+
+			feeAddress, err := sdk.AccAddressFromBech32(channelFee.FeeAddress)
+			if err != nil {
+				return nil, err
+			}
+
+			k.bank.SendCoins(ctx, msgSender, feeAddress, sdk.NewCoins(sdk.NewCoin(msg.Token.Denom, charge)))
+
+			if newAmount.IsZero() {
```

Comment:
when will this happen?

can you add more docs to the flow in general?

Reply:
This could happen if min fee is X but the transfer amount is also X. so there is not possible to charge the percentage fee but transfer only min fee to the fee collector account and ignore IBC

Reply:
Added docs under the method that explain all behaviour.